src/Symfony/Component/Config/Definition/Builder/NodeDefinition.php

```diff
@@ -22,21 +22,21 @@
  */
 abstract class NodeDefinition implements NodeParentInterface
 {
-    protected $name;
-    protected $normalization;
-    protected $validation;
-    protected $defaultValue;
-    protected $default = false;
-    protected $required = false;
-    protected $deprecation = [];
-    protected $merge;
-    protected $allowEmptyValue = true;
-    protected $nullEquivalent;
-    protected $trueEquivalent = true;
-    protected $falseEquivalent = false;
-    protected $pathSeparator = BaseNode::DEFAULT_PATH_SEPARATOR;
-    protected $parent;
-    protected $attributes = [];
+    protected ?string $name;
+    protected NormalizationBuilder $normalization;
+    protected ValidationBuilder $validation;
+    protected mixed $defaultValue;
+    protected bool $default = false;
+    protected bool $required = false;
+    protected array $deprecation = [];
+    protected MergeBuilder $merge;
+    protected bool $allowEmptyValue = true;
+    protected mixed $nullEquivalent = null;
+    protected mixed $trueEquivalent = true;
+    protected mixed $falseEquivalent = false;
+    protected string $pathSeparator = BaseNode::DEFAULT_PATH_SEPARATOR;
+    protected NodeParentInterface|NodeInterface|null $parent;
```

Comment:
All those classes in the return type of `end()` implement `NodeParentInterface` so we could reduce this to only it. I suppose the list is here to hint the IDE for autocompletion. Maybe we should keep only `NodeParentInterface` as native type and use `@return` for the rest?

The other strange thing is about the need to add `NodeInterface` to make this code work:
https://github.com/symfony/symfony/blob/1935af612953715faec8b69f2be70b20c7427bc5/src/Symfony/Component/Config/Definition/Builder/ArrayNodeDefinition.php#L352-L361

`end()` can't return an `ArrayNode` because of its return type. Either the return type is wrong, or what?
Up to investigate and fix this on a lower branch?

Reply:
> All those classes in the return type of `end()` implement `NodeParentInterface` so we could reduce this to only it. I suppose the list is here to hint the IDE for autocompletion. Maybe we should keep only `NodeParentInterface` as native type and use `@return` for the rest?

Would work for me to unblock the PR.

> Up to investigate and fix this on a lower branch?

Maybe, but not right away. 🙈